### ISSUE TITLE
Calls to cast_types using torch.nn.Parameter.data evade autocast, leading to problems with mixed precision

### ISSUE THREAD

**Issue by reppertj** [KevinMusgrave/pytorch-metric-learning#293]
First of all, thanks for this great library; it's saved me tons of time in experimenting with different metric learning models and metrics.

I've run into several issues trying to use some functions with automatic mixed precision, and I believe I've tracked down the cause. As an example, `compute_loss` in ProxyAnchorLoss currently tries to ensure the compatibility of the embeddings and the proxies by reassigning self.proxies.data to match the dtype and device of the embeddings. However, since autocast respects the `dtype=...` argument (and, I think, doesn't track assignments directly to `Tensor.data` in any case), the proxies end up having the wrong dtype after leaving the autocast context, because the embeddings come in as FP16 although they are "really" FP32. This causes unexpected results downstream in mixed precision training using AMP. This is possibly related to issue #228, but I didn't verify this.

I'm not sure how best to address this. I have subclassed the loss function and removed the `cast_types` call in `compute_loss`, which remedied the problem for me. Recasting the proxies back to FP32 after leaving an autocast enabled context also works, but it's probably less than ideal, as ops done between the FP16 cast and the FP32 cast might not benefit from AMP. I'm not sure whether there is a safe way to fix this in general without breaking the current API. In general, choosing the dtype and device of an nn.Module and its parameters is something I expect to do as a user, and not have managed for me inside the module. In a mixed precision context, I don't think it's safe to change the dtype of a parameter based on the dtype of the inputs (i.e., `dtype=...` should be avoided unless you really mean it).

Sorry I don't have an obvious fix! Just thought I would flag this, though, if at least to add a note to the documentation, as this took me a while to track down. I'm happy to provide more details or do some more digging if it would help.

**Comment by KevinMusgrave** [comment]
Thanks for bringing this up. I'd like to have this working before the sacred v1.0 is released, so I'll break whatever is necessary. 

If autocast is enabled then the tensors created by a loss function should never be manually typecast. But if autocast is not enabled, then those tensors **do** need to be typecast.

So should loss functions come with an ```autocast``` flag? If set to True, then no manual typecasting would be done.

**Comment by KevinMusgrave** [comment]
Actually a flag might be unnecessary. It looks like there is a ```torch.is_autocast_enabled()``` function.

**Comment by KevinMusgrave** [comment]
The code change might be as simple as adding an ```if torch.is_autocast_enabled():``` to the beginning of this function:

https://github.com/KevinMusgrave/pytorch-metric-learning/blob/251bde4cb85d8011141bb768c3b355e925713c10/src/pytorch_metric_learning/utils/common_functions.py#L460-L464

**Comment by reppertj** [comment]
Yes, I think that should do it! I didn't know about `torch.is_autocast_enabled()` — that's actually helpful to have.

**Comment by KevinMusgrave** [comment]
I made the change:
```
pip install pytorch-metric-learning==0.9.98.dev1
```
If you try it, let me know if it works.

**Comment by reppertj** [comment]
I just tried ProxyAnchorLoss from the dev branch with autocast, and it looks good! (Now, failing to converge is my problem 😄)

**Comment by KevinMusgrave** [comment]
Going to keep this open until its in the master branch
